File: app/src/main/java/com/example/joseramos/lyricsappjlrf/presentation/activities/base/BaseActivity.java
package com.example.joseramos.lyricsappjlrf.presentation.activities.base;


import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

import com.example.joseramos.lyricsappjlrf.MyLyricsApplication;
import com.example.joseramos.lyricsappjlrf.di.ComponentFactory;
import com.example.joseramos.lyricsappjlrf.di.components.ActivityComponent;
import com.example.joseramos.lyricsappjlrf.presentation.navigation.Navigator;
import com.example.joseramos.lyricsappjlrf.presentation.navigation.NavigatorManager;
import com.example.joseramos.lyricsappjlrf.presentation.navigation.OnBackPressListener;

import javax.inject.Inject;

public abstract class BaseActivity extends AppCompatActivity {

    private ActivityComponent activityComponent;

    @Inject
    protected NavigatorManager navigatorManager;

    @Inject
    protected Navigator navigator;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activityComponent = ComponentFactory.createActivityComponent(
                MyLyricsApplication.getComponent(this),
                getSupportFragmentManager(),
                this);
        activityComponent.inject(this);
    }

    @Override
    public void onBackPressed() {
        if (navigatorManager.getCurrentFragment() != null && navigatorManager.getCurrentFragment() instanceof OnBackPressListener) {
            ((OnBackPressListener) navigatorManager.getCurrentFragment()).onBackPressed();
        } else {
            super.onBackPressed();
            navigatorManager.popStack();
        }
    }

}

File: app/src/main/java/com/example/joseramos/lyricsappjlrf/presentation/activities/base/BaseActivity.kt
package com.example.joseramos.lyricsappjlrf.presentation.activities.base

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import com.example.joseramos.lyricsappjlrf.presentation.navigation.Navigator
import com.example.joseramos.lyricsappjlrf.presentation.navigation.OnBackPressListener
import dagger.android.AndroidInjection
import javax.inject.Inject

abstract class BaseActivity : AppCompatActivity() {
    @Inject
    lateinit var navigator: Navigator

    override fun onCreate(savedInstanceState: Bundle?) {
        //FIXME: Remove when AppInjector is used to inject any android component ( For Example Activities and Fragments).
        AndroidInjection.inject(this)
        super.onCreate(savedInstanceState)
    }

    override fun onBackPressed() {
        if (navigator.getNavigatorManager().currentFragment != null && navigator.getNavigatorManager().currentFragment is OnBackPressListener) {
            (navigator.getNavigatorManager().currentFragment as OnBackPressListener).onBackPressed()
        } else {
            super.onBackPressed()
            navigator.getNavigatorManager().popStack()
        }
    }
}
